pkg/audit/cursor.go

```diff
@@ -120,19 +91,72 @@ func getRandomStripe(pointer *pb.Pointer) (index int64, err error) {
 		return 0, nil
 	}
 
-	randomStripeIndex, err := rand.Int(rand.Reader, big.NewInt(pointer.GetSegmentSize()/int64(redundancy.StripeSize())))
-	if err != nil {
-		return -1, err
+	var src cryptoSource
+	rnd := rand.New(src)
+	numStripes := pointer.GetSegmentSize() / int64(redundancy.StripeSize())
+	randomStripeIndex := rnd.Int63n(numStripes)
+
+	return randomStripeIndex, nil
+}
+
+// getRandomValidPointer attempts to get a random remote pointer from a list. If it sees expired pointers in the process of looking, deletes them
+func (cursor *Cursor) getRandomValidPointer(pointerItems []*pb.ListResponse_Item) (pointer *pb.Pointer, path storj.Path, err error) {
+	var src cryptoSource
+	rnd := rand.New(src)
+	errGroup := new(errs.Group)
+	randomNums := rnd.Perm(len(pointerItems))
+
+	for _, randomIndex := range randomNums {
+		pointerItem := pointerItems[randomIndex]
+		path := pointerItem.Path
+
+		// get pointer info
+		pointer, err := cursor.metainfo.Get(path)
+		if err != nil {
+			errGroup.Add(err)
+			continue
+		}
+
+		//delete expired items rather than auditing them
+		if expiration := pointer.GetExpirationDate(); expiration != nil {
+			t, err := ptypes.Timestamp(expiration)
+			if err != nil {
+				errGroup.Add(err)
+				continue
+			}
+			if t.Before(time.Now()) {
+				err := cursor.metainfo.Delete(path)
+				if err != nil {
+					errGroup.Add(err)
+				}
+				continue
+			}
+		}
+
+		if pointer.GetType() != pb.Pointer_REMOTE || pointer.GetSegmentSize() == 0 {
+			continue
+		}
+
+		return pointer, path, nil
 	}
 
-	return randomStripeIndex.Int64(), nil
+	errGroup.Add(Error.New("no valid node found in selection"))
+	return nil, "", errGroup.Err()
+}
+
+// cryptoSource implements the math/rand Source interface using crypto/rand
+type cryptoSource struct{}
+
+func (s cryptoSource) Seed(seed int64) {}
```

Comment:
Do we expect any call to this or if there is a call to this we should say "don't do this call"?

Reply:
never mind, I've just realized that could be called by `rand.New`